rest-api/dao/biobank_stored_sample_dao.py

```diff
@@ -13,17 +14,32 @@ def get_id(self, obj):
     return obj.biobankStoredSampleId
 
   def upsert_all(self, sample_generator):
-    """Inserts/updates samples. Raises ValueError for invalid samples."""
+    """Inserts/updates samples. Raises ValueError for invalid samples and re-attempts in case of a
+    database connection lost."""
     # TODO(DA-230) Scale this to handle full study data.
     # SQLAlchemy does not provide batch upserting; individual session.merge() calls as below may be
     # expensive but cannot be effectively batched, see stackoverflow.com/questions/25955200. If this
     # proves to be a bottleneck, we can switch to generating "INSERT .. ON DUPLICATE KEY UPDATE".
-    written = 0
-    with self.session() as session:
+    tries = 0
+    while tries < 10:
+      written = 0
+      session = self._database.make_session()
       for sample in sample_generator:
         if sample.test not in BIOBANK_TESTS_SET:
           logging.warn('test sample %s not recognized.' % sample.test)
         else:
           session.merge(sample)
           written += 1
-    return written
+      try:
+        session.commit()
+      except DBAPIError as exc:
```

Comment:
We could consider having a function that does the loop, try, catch, etc. and calls a function passed in as a parameter to perform the actual logic; or you could use a decorator for this purpose. That would make it easier to use this sort of logic elsewhere.

Reply:
This is a good idea but I think it merits further thought.  Making a function that wraps session actually shouldn't be that hard (and if we use something like `tenacity` making it retry shouldn't be that hard either), the trickier question is knowing all the places we need to rewrite to use that function.  If we wanted to use it at the DAO level, for example, there are 8 places in `BaseDao` where `with self.session()` is used, and all 8 would need to be converted from using the context manager to using a function that takes a function.  I vote that we leave this Jira ticket open but merge this PR and make further work subject to a new PR.

Reply:
Yeah, I wasn't necessarily suggesting we actually use this in other places yet... just that we write the code we can use in future whenever we spot another place where we'd like to retry.

Reply:
A couple places that we may want to consider reusing this kind of logic would be when writing metrics:

https://github.com/all-of-us/raw-data-repository/blob/master/rest-api/offline/metrics_pipeline.py#L631
https://github.com/all-of-us/raw-data-repository/blob/master/rest-api/offline/public_metrics_export.py#L227